In [1]:
import numpy as np
import pylab as pl
import pandas as pd
from sklearn.cluster import KMeans 
from sklearn import preprocessing

np.set_printoptions(precision=2, linewidth=500, suppress=True, threshold=np.nan)

from IPython.display import display
pd.options.display.max_columns = None
pd.options.display.max_rows = None

In [2]:
## 1.a load data

## Rows = images
## Columns = features

data = pd.read_csv('segmentation_data.txt', header=None)
classes = pd.read_table('segmentation_classes.txt', sep='\t', header=None)

In [3]:
data.head(10)

0      1   2         3    4         5         6         7         8   \
0  110.0  189.0   9  0.000000  0.0  1.000000  0.666667  1.222222  1.186342   
1   86.0  187.0   9  0.000000  0.0  1.111111  0.720082  1.444444  0.750309   
2  225.0  244.0   9  0.000000  0.0  3.388889  2.195113  3.000000  1.520234   
3   47.0  232.0   9  0.000000  0.0  1.277778  1.254621  1.000000  0.894427   
4   97.0  186.0   9  0.000000  0.0  1.166667  0.691215  1.166667  1.005540   
5  157.0  221.0   9  0.000000  0.0  1.055556  0.646930  1.222222  0.620633   
6   62.0  224.0   9  0.000000  0.0  0.944445  1.083547  2.333333  1.632993   
7   27.0  248.0   9  0.111111  0.0  1.611111  0.646930  3.166667  1.722401   
8   44.0  233.0   9  0.000000  0.0  2.222222  2.146487  2.111111  1.327766   
9   17.0  229.0   9  0.000000  0.0  2.111111  1.985130  2.444445  1.614747   

          9          10         11         12        13         14         15  \
0  12.925926  10.888889   9.222222  18.666668 -6.111111 -11.111111  17.222221   
1  13.740741  11.666667  10.333334  19.222221 -6.222222 -10.222222  16.444445   
2  12.259259  10.333334   9.333334  17.111110 -5.777778  -8.777778  14.555555   
3  12.703704  11.000000   9.000000  18.111110 -5.111111 -11.111111  16.222221   
4  15.592592  13.888889  11.777778  21.111110 -5.111111 -11.444445  16.555555   
5  12.111111  10.222222   8.111112  18.000000 -5.666666 -12.000000  17.666666   
6  14.629630  13.222222  11.444445  19.222221 -4.222222  -9.555555  13.777778   
7  15.296296  14.777778  12.888889  18.222221 -1.555556  -7.222222   8.777778   
8  14.481482  12.555555  11.333333  19.555555 -5.777778  -9.444445  15.222222   
9  13.703704  11.222222  11.111111  18.777779 -7.444445  -7.777778  15.222222   

          16        17        18  
0  18.666668  0.508139  1.910864  
1  19.222221  0.463329  1.941465  
2  17.111110  0.480149  1.987902  
3  18.111110  0.500966  1.875362  
4  21.111110  0.442661  1.863654  
5  18.000000  0.549180  1.877146  
6  19.222221  0.408965  1.860191  
7  18.222221  0.312227  1.783512  
8  19.555555  0.422174  1.950405  
9  18.777779  0.439852  2.099904

In [5]:
classes.head(10)

0  1
0  GRASS  0
1  GRASS  0
2  GRASS  0
3  GRASS  0
4  GRASS  0
5  GRASS  0
6  GRASS  0
7  GRASS  0
8  GRASS  0
9  GRASS  0

In [6]:
## 1.a Perform Min-Max Normalization

min_max_scaler = preprocessing.MinMaxScaler().fit(data)
data_norm = min_max_scaler.transform(data)

data_norm[0:5]

array([[0.43, 0.74, 0.  , 0.  , 0.  , 0.03, 0.  , 0.03, 0.  , 0.09, 0.08, 0.06, 0.13, 0.73, 0.01, 0.87, 0.12, 0.51, 0.83],
       [0.34, 0.73, 0.  , 0.  , 0.  , 0.04, 0.  , 0.03, 0.  , 0.1 , 0.09, 0.07, 0.13, 0.73, 0.02, 0.86, 0.13, 0.46, 0.84],
       [0.89, 0.97, 0.  , 0.  , 0.  , 0.12, 0.  , 0.07, 0.  , 0.09, 0.08, 0.06, 0.12, 0.74, 0.04, 0.83, 0.11, 0.48, 0.84],
       [0.18, 0.92, 0.  , 0.  , 0.  , 0.04, 0.  , 0.02, 0.  , 0.09, 0.08, 0.06, 0.13, 0.75, 0.01, 0.86, 0.12, 0.5 , 0.83],
       [0.38, 0.73, 0.  , 0.  , 0.  , 0.04, 0.  , 0.03, 0.  , 0.11, 0.1 , 0.08, 0.15, 0.75, 0.01, 0.86, 0.14, 0.44, 0.82]])

In [7]:
## 1.b perfrom Kmeans clustering

kmeans = KMeans(n_clusters=7, max_iter=500, verbose=1) # initialization
kmeans.fit(data_norm)

Initialization complete
start iteration
done sorting
end inner loop
Iteration 0, inertia 409.8535598348767
start iteration
done sorting
end inner loop
Iteration 1, inertia 372.47496521049527
start iteration
done sorting
end inner loop
Iteration 2, inertia 366.72435039734194
start iteration
done sorting
end inner loop
Iteration 3, inertia 363.9609802349181
start iteration
done sorting
end inner loop
Iteration 4, inertia 361.52469166838796
start iteration
done sorting
end inner loop
Iteration 5, inertia 358.7053423609243
start iteration
done sorting
end inner loop
Iteration 6, inertia 355.45825396408077
start iteration
done sorting
end inner loop
Iteration 7, inertia 352.8574622907119
start iteration
done sorting
end inner loop
Iteration 8, inertia 351.6331682657651
start iteration
done sorting
end inner loop
Iteration 9, inertia 351.32180554867097
start iteration
done sorting
end inner loop
Iteration 10, inertia 351.08070725620126
start iteration
done sorting
end inner loop
Iteration 11

Iteration 7, inertia 368.83866892064407
start iteration
done sorting
end inner loop
Iteration 8, inertia 368.81624315452945
start iteration
done sorting
end inner loop
Iteration 9, inertia 368.8079421756174
start iteration
done sorting
end inner loop
Iteration 10, inertia 368.8079421756174
center shift 0.000000e+00 within tolerance 4.150157e-06
Initialization complete
start iteration
done sorting
end inner loop
Iteration 0, inertia 382.01103618006556
start iteration
done sorting
end inner loop
Iteration 1, inertia 373.915220384805
start iteration
done sorting
end inner loop
Iteration 2, inertia 371.4466219918702
start iteration
done sorting
end inner loop
Iteration 3, inertia 370.442358515848
start iteration
done sorting
end inner loop
Iteration 4, inertia 369.97001920739393
start iteration
done sorting
end inner loop
Iteration 5, inertia 369.8179288550935
start iteration
done sorting
end inner loop
Iteration 6, inertia 369.7483254655608
start iteration
done sorting
end inner loop
Iter

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=500,
    n_clusters=7, n_init=10, n_jobs=1, precompute_distances='auto',
    random_state=None, tol=0.0001, verbose=1)

In [8]:
clusters = kmeans.predict(data_norm)

In [9]:
clusters

array([2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 0, 4, 0, 0, 0, 0, 4, 4, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0, 4, 0, 4, 0, 4, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 4, 0, 0, 0, 0, 4, 0, 4, 0, 0, 0, 4, 4, 4, 0, 4, 0, 0, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5,
       5, 5, 3, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 4, 0, 0, 4, 0, 4, 4, 0, 5, 0, 4, 0, 4, 0, 4, 4, 4, 6, 4, 4, 0, 0, 4, 4, 4, 4, 0, 4, 0, 0, 4, 0, 0, 0, 4, 4, 4, 0, 6, 0, 4, 4, 4, 4, 4, 0, 0, 0, 0, 0, 6, 4, 4, 4, 6, 4, 0, 4, 0, 0, 4, 4, 4, 0, 4, 6, 0, 4, 4, 4, 4, 4, 4, 4, 4, 4, 0, 0, 0, 4, 0, 0, 4, 0, 6, 4, 4, 0, 4, 4, 4, 0, 4, 6, 0,
  

In [10]:
centroids = kmeans.cluster_centers_

In [11]:
# 1.b print centroids

centroids

array([[0.75, 0.53, 0.  , 0.04, 0.04, 0.11, 0.02, 0.11, 0.02, 0.3 , 0.28, 0.35, 0.26, 0.59, 0.45, 0.31, 0.35, 0.3 , 0.16],
       [0.54, 0.15, 0.  , 0.03, 0.  , 0.03, 0.  , 0.03, 0.  , 0.82, 0.78, 0.89, 0.79, 0.27, 0.67, 0.29, 0.89, 0.21, 0.13],
       [0.51, 0.81, 0.  , 0.08, 0.01, 0.05, 0.  , 0.05, 0.  , 0.11, 0.09, 0.09, 0.14, 0.68, 0.08, 0.82, 0.13, 0.41, 0.89],
       [0.25, 0.46, 0.  , 0.03, 0.01, 0.04, 0.  , 0.03, 0.  , 0.03, 0.02, 0.04, 0.02, 0.77, 0.22, 0.51, 0.04, 0.8 , 0.18],
       [0.3 , 0.53, 0.  , 0.05, 0.05, 0.1 , 0.01, 0.08, 0.01, 0.4 , 0.37, 0.47, 0.35, 0.5 , 0.57, 0.21, 0.47, 0.3 , 0.16],
       [0.77, 0.43, 0.  , 0.01, 0.02, 0.04, 0.  , 0.02, 0.  , 0.04, 0.03, 0.06, 0.03, 0.78, 0.22, 0.49, 0.06, 0.54, 0.24],
       [0.25, 0.39, 0.  , 0.08, 0.02, 0.08, 0.  , 0.06, 0.01, 0.15, 0.14, 0.18, 0.12, 0.72, 0.34, 0.35, 0.18, 0.41, 0.2 ]])

In [12]:
# 1.b completeness and homogeneity

from sklearn.metrics import completeness_score, homogeneity_score

In [13]:
print completeness_score(classes[1],clusters)

0.6116744999910891


In [14]:
print homogeneity_score(classes[1],clusters)

0.6099656393147241


In [15]:
# 1.c Perform PCA

from sklearn import decomposition
pca = decomposition.PCA(n_components=6)

data_trans = pca.fit(data_norm).transform(data_norm)
print data_trans

[[-0.69  0.53  0.25 -0.2  -0.08  0.05]
 [-0.67  0.51  0.34 -0.17 -0.04  0.06]
 [-0.71  0.77 -0.16 -0.01 -0.17  0.04]
 [-0.73  0.51  0.5  -0.06 -0.14  0.03]
 [-0.64  0.53  0.3  -0.18 -0.02  0.05]
 [-0.73  0.63  0.08 -0.14 -0.17  0.02]
 [-0.67  0.53  0.44 -0.02 -0.06  0.03]
 [-0.66  0.52  0.59  0.15  0.01 -0.17]
 [-0.69  0.53  0.52 -0.01 -0.1   0.05]
 [-0.69  0.5   0.62 -0.02 -0.12  0.06]
 [-0.73  0.52  0.56 -0.01 -0.15  0.04]
 [-0.68  0.51  0.47 -0.04 -0.02 -0.15]
 [-0.72  0.59  0.15 -0.16 -0.14  0.05]
 [-0.69  0.52  0.32 -0.19 -0.08  0.05]
 [-0.77  0.77 -0.23 -0.07 -0.23 -0.19]
 [-0.72  0.5   0.45 -0.08 -0.12  0.04]
 [-0.74  0.57  0.11 -0.19 -0.14  0.02]
 [-0.66  0.42  0.58 -0.15 -0.    0.05]
 [-0.53  0.79 -0.23 -0.18  0.02  0.05]
 [-0.61  0.74 -0.16 -0.22 -0.01 -0.15]
 [-0.71  0.52  0.37 -0.15 -0.12  0.06]
 [-0.68  0.66 -0.02 -0.16 -0.13  0.09]
 [-0.67  0.59  0.38 -0.04 -0.07 -0.15]
 [-0.7   0.71  0.1  -0.03 -0.14 -0.15]
 [-0.64  0.69  0.27  0.03 -0.12  0.05]
 [-0.62  0.48  0.5  -0.1 

In [16]:
# six components explain 95% of the variance

print(pca.explained_variance_ratio_)

[0.61 0.13 0.1  0.05 0.04 0.02]


In [17]:
# 1.c Perform kmeans on reduced dimensions

kmeans = KMeans(n_clusters=7, max_iter=500, verbose=1) # initialization
kmeans.fit(data_trans)

Initialization complete
start iteration
done sorting
end inner loop
Iteration 0, inertia 307.2470886508099
start iteration
done sorting
end inner loop
Iteration 1, inertia 283.99037171519586
start iteration
done sorting
end inner loop
Iteration 2, inertia 281.4324513291552
start iteration
done sorting
end inner loop
Iteration 3, inertia 280.1048863263177
start iteration
done sorting
end inner loop
Iteration 4, inertia 278.5706740260834
start iteration
done sorting
end inner loop
Iteration 5, inertia 276.11833789259174
start iteration
done sorting
end inner loop
Iteration 6, inertia 274.07674639877547
start iteration
done sorting
end inner loop
Iteration 7, inertia 272.57756940008113
start iteration
done sorting
end inner loop
Iteration 8, inertia 271.6337065624085
start iteration
done sorting
end inner loop
Iteration 9, inertia 271.16034014981875
start iteration
done sorting
end inner loop
Iteration 10, inertia 271.01512530715024
start iteration
done sorting
end inner loop
Iteration 11

Iteration 33, inertia 276.9378791654419
start iteration
done sorting
end inner loop
Iteration 34, inertia 276.9121150856288
start iteration
done sorting
end inner loop
Iteration 35, inertia 276.90678848789935
center shift 3.005913e-03 within tolerance 1.236861e-05
Initialization complete
start iteration
done sorting
end inner loop
Iteration 0, inertia 287.5473399157353
start iteration
done sorting
end inner loop
Iteration 1, inertia 272.30150809106965
start iteration
done sorting
end inner loop
Iteration 2, inertia 269.42868435088405
start iteration
done sorting
end inner loop
Iteration 3, inertia 267.46546546085995
start iteration
done sorting
end inner loop
Iteration 4, inertia 263.88806044276896
start iteration
done sorting
end inner loop
Iteration 5, inertia 261.88059982421794
start iteration
done sorting
end inner loop
Iteration 6, inertia 260.62291518360144
start iteration
done sorting
end inner loop
Iteration 7, inertia 259.72660564901713
start iteration
done sorting
end inner l

KMeans(algorithm='auto', copy_x=True, init='k-means++', max_iter=500,
    n_clusters=7, n_init=10, n_jobs=1, precompute_distances='auto',
    random_state=None, tol=0.0001, verbose=1)

In [18]:
clusters_pca = kmeans.predict(data_trans)
clusters_pca

array([4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 4, 5, 2, 5, 5, 5, 5, 2, 2, 5, 5, 5, 5, 5, 5, 5, 2, 5, 5, 5, 5, 2, 5, 2, 5, 2, 5, 5, 5, 5, 5, 2, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 2, 5, 2, 5, 5, 5, 5, 5, 5, 5, 5, 5, 5, 2, 5, 5, 5, 5, 5, 2, 5, 5, 5, 5, 5, 5, 5, 2, 5, 5, 2, 5, 5, 5, 5, 2, 5, 2, 5, 5, 5, 2, 2, 2, 5, 2, 5, 5, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3,
       3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 6, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 3, 2, 5, 5, 2, 5, 2, 2, 5, 3, 5, 2, 5, 2, 5, 2, 2, 2, 6, 2, 2, 5, 5, 2, 2, 2, 2, 5, 2, 5, 5, 2, 5, 5, 5, 2, 2, 2, 5, 6, 5, 2, 2, 2, 2, 2, 5, 5, 5, 5, 5, 6, 2, 2, 2, 6, 2, 5, 2, 5, 5, 2, 2, 2, 5, 2, 6, 5, 2, 2, 2, 2, 2, 2, 2, 2, 2, 5, 5, 5, 2, 5, 5, 2, 5, 6, 2, 2, 5, 2, 2, 2, 5, 2, 6, 5,
  

In [19]:
centroids_pca = kmeans.cluster_centers_
centroids_pca

array([[-0.6 , -0.36,  0.11, -0.13, -0.13, -0.02],
       [ 1.41,  0.09,  0.04, -0.17, -0.03, -0.01],
       [ 0.44, -0.1 ,  0.17,  0.24, -0.05, -0.01],
       [-0.51, -0.07, -0.34, -0.07,  0.08,  0.01],
       [-0.62,  0.64,  0.2 , -0.09, -0.07,  0.01],
       [ 0.18,  0.05, -0.27,  0.19,  0.03,  0.02],
       [-0.2 , -0.24,  0.15,  0.06,  0.13, -0.01]])

In [20]:
print completeness_score(classes[1],clusters_pca)

0.6108391577738163


In [21]:
print homogeneity_score(classes[1],clusters_pca)

0.6091689230222487


### 1.e discuss

#### Scores for full data

Completeness Score = 0.6116744999910891 <br>
Homogeneity Score = 0.6099656393147241 <br>

#### Scores after PCA

Completeness Score = 0.6108391577738163 <br>
Homogeneity Score = 0.6091689230222487 <br>

The scores before and after PCA are essentially the same. This illustrates that the performance of a model can remain unchanged when utilizing PCA to reduce dimensionality. PCA only retains those features that provide value to the model.

In [23]:
# 2

jokes = pd.read_csv('jokes.csv', header=None)
ratings = pd.read_csv('modified_jester_data.csv', header=None)

In [24]:
jokes.head(10)

0                                                  1
0  0  A man visits the doctor. The doctor says "I ha...
1  1  This couple had an excellent relationship goin...
2  2  Q. What's 200 feet long and has 4 teeth? A. Th...
3  3  Q. What's the difference between a man and a t...
4  4  Q. What's O. J. Simpson's Internet address? A....
5  5  Bill & Hillary are on a trip back to Arkansas....
6  6  How many feminists does it take to screw in a ...
7  7  Q. Did you hear about the dyslexic devil worsh...
8  8  A country guy goes into a city bar that has a ...
9  9  Two cannibals are eating a clown one turns to ...

In [25]:
ratings.head(10)

0      1      2      3      4      5      6      7      8      9   \
0   3.18  19.79   1.34   2.84   3.48   2.50   1.15  15.17   2.02   6.24   
1  15.08  10.71  17.36  15.37   8.62   1.34  10.27   5.66  19.88  20.22   
2   0.00   0.00   0.00   0.00  20.03  20.27  20.03  20.27   0.00   0.00   
3   0.00  19.35   0.00   0.00  12.80  19.16   8.18  17.21   0.00  12.84   
4  19.50  15.61   6.83   5.61  12.36  12.60  18.04  15.61  10.56  16.73   
5   4.83   7.46  11.44   2.50   3.91   6.68   2.31  10.13   4.35   9.20   
6   0.00   0.00   0.00   0.00  19.59   1.15  18.72  19.79   0.00   0.00   
7  17.84  14.16  20.17   4.79   2.84   9.30  20.27  12.41   5.81   6.58   
8   7.21   7.46   1.58   4.11   2.26  10.71   5.71   2.07   3.14   9.40   
9  14.01  16.15  16.15  14.01  17.41  16.15  19.93  13.52  14.01  19.16   

      10     11     12     13     14     15     16     17     18     19  \
0   2.50   4.25   3.82  19.45   3.82   3.48   3.57   1.19   1.15   1.15   
1  17.75  19.64  15.42  18.43  15.56  10.03  15.66  10.32  14.30   9.79   
2  18.33  18.57  20.37  17.17   4.64   4.11   3.14  20.03  20.03  20.03   
3  18.33  17.60  17.31  19.11   3.77   4.35  12.17   4.40   7.36   8.91   
4  19.25  17.84   7.07  18.23   8.67   1.34  13.72   9.64  13.57  15.51   
5   4.20   5.27   6.00   2.41  11.49   2.07   7.31   8.82   8.72   4.88   
6  15.27  18.62   4.74  13.96  17.07   7.50   8.91  17.17  16.15  15.42   
7  19.20   3.14   4.06   3.04  11.29   1.10   3.91   3.82  12.02  10.71   
8   8.09  10.71   6.15  10.51   2.26   4.01   2.26   8.09   7.65  10.71   
9  16.53  17.02  15.47  16.44   6.34  10.03  10.56  12.55  11.49  15.37   

      20     21     22     23     24     25     26     27     28     29  \
0   1.63  12.50   6.63   1.19   2.50  12.12  18.82  13.86  20.13   3.57   
1  11.87  19.64  19.35  20.17  11.05  18.57  15.71  11.87  10.61  17.99   
2  18.28   0.00  19.25   0.00   0.00  18.48  18.28  18.28  19.93   0.00   
3  16.34   0.00   0.00   0.00   0.00  13.91  14.93  17.75  17.60   0.00   
4  19.20  17.12  19.30   9.74  18.77  12.89   9.83  16.68  19.45  15.61   
5   7.99  10.42   8.62   3.23   8.77   8.72   5.76   5.47  10.32   4.06   
6  16.63  13.43   0.00   0.00  15.13  18.09  18.33  18.18  11.92   0.00   
7   6.29   3.57  12.12   2.75  14.79  11.10  14.45   7.55  18.57   6.83   
8  14.98   9.40  10.71  12.21  14.79  13.91  15.85  13.28  11.97  10.71   
9  16.44  16.24  17.60  14.01  17.02  17.80  13.91  16.15  14.50  16.15   

      30     31     32     33     34     35     36     37     38     39  \
0  13.14   6.92   1.92  18.82  16.05  15.95   1.83   2.60   2.60   2.60   
1  17.50  10.08  18.14  20.03   9.20  11.73  18.09  14.40  10.13  18.91   
2  17.17  18.28   0.00   0.00  19.98  18.33   0.00  17.17  20.08  18.33   
3  17.65   4.88   0.00  18.57  17.21  17.65   0.00   2.70  18.18  13.82   
4  19.06   1.53  18.28  16.68  13.48  14.20   9.74  17.80  15.51  13.48   
5   7.07  10.66   4.40   5.51   9.45  10.51   4.79   2.94   5.61   3.96   
6  16.15  16.87   0.00   0.00  18.96  14.79   0.00  12.55   0.00  14.11   
7  10.56  20.27  16.83   7.60   5.56  12.60   3.86  14.01   4.93  16.68   
8  17.36  11.53  10.71  15.66  14.35  11.97   7.89   9.16  15.42   6.58   
9  17.60  10.90  14.11  17.50  20.22  18.86  14.01  17.50  14.01  18.67   

      40     41     42     43     44     45     46     47     48     49  \
0   2.89   1.87   1.97   1.92   3.86   4.74  14.79  10.90  14.93  15.13   
1  18.82   7.17  19.64  19.98   6.68   7.80   6.10  10.08   9.54  14.64   
2  18.52  20.27  20.27   0.00  20.27   0.00  17.17  18.33  20.08  18.28   
3   0.00  12.55   0.00   0.00   0.00  17.84  17.84   7.02  17.99  14.45   
4  11.34  17.84  11.19   2.26  16.24  17.31  19.06  12.26  15.51  10.95   
5   2.17   3.09  12.07   9.11   4.64   6.78   1.73   1.19   3.18  10.37   
6  17.26   0.00   0.00   0.00  17.99  14.20   0.00  18.33  15.71  14.54   
7   4.54   6.58   2.02  11.53   9.74   6.58   2.07  19.93  10.90  18.43   
8   9.59  12.41   5.66   4.

In [29]:
for row in ratings:
    print ratings[row]

0       3.18
1      15.08
2       0.00
3       0.00
4      19.50
5       4.83
6       0.00
7      17.84
8       7.21
9      14.01
10      8.09
11     12.31
12      0.00
13     20.22
14     19.79
15      7.50
16      0.00
17     14.16
18     15.22
19      0.00
20     13.57
21     18.28
22     13.72
23     13.09
24      7.07
25      0.00
26      0.00
27     11.49
28      0.00
29      0.00
30     12.94
31      0.00
32     12.02
33      0.00
34     16.92
35      1.34
36     18.18
37      7.31
38     13.96
39      0.00
40      6.05
41      0.00
42      0.00
43     11.15
44     19.79
45     18.33
46     19.45
47      2.75
48      5.56
49      3.18
50     14.16
51      0.00
52      0.00
53      2.41
54     19.11
55      8.72
56     17.21
57     17.94
58      3.57
59      0.00
60     15.56
61     12.26
62     11.29
63     12.26
64     10.71
65      0.00
66      9.40
67     14.40
68      3.96
69      0.00
70     14.50
71      4.79
72     16.44
73     20.27
74     13.38
75      0.00
76      2.50

0       1.15
1      10.27
2      20.03
3       8.18
4      18.04
5       2.31
6      18.72
7      20.27
8       5.71
9      19.93
10     13.48
11     10.03
12     16.58
13     14.50
14     19.69
15      8.04
16     19.74
17     13.43
18     15.13
19     19.93
20      6.78
21     19.93
22     18.04
23      5.81
24      7.07
25     11.10
26      6.53
27     11.29
28     19.83
29     11.44
30     14.83
31     15.81
32     11.97
33     10.81
34     13.72
35      1.78
36      4.88
37     13.18
38     14.74
39      7.65
40     10.03
41     13.72
42      6.49
43     13.04
44     15.51
45      2.84
46     19.59
47     10.47
48     14.30
49     18.77
50      6.87
51     13.18
52      9.98
53      1.92
54     16.78
55     13.33
56     18.18
57     13.67
58      3.14
59     17.70
60     12.65
61     12.84
62      1.83
63      6.29
64      4.11
65     20.13
66     11.00
67     11.73
68      4.88
69     13.67
70      8.18
71      3.96
72      4.20
73     18.91
74      1.39
75     12.75
76      2.55

0       3.82
1      15.56
2       4.64
3       3.77
4       8.67
5      11.49
6      17.07
7      11.29
8       2.26
9       6.34
10     11.92
11      9.20
12      6.15
13     12.07
14      7.21
15     13.52
16     18.72
17     13.57
18     12.65
19      9.98
20      5.90
21     10.32
22      4.35
23      3.86
24     10.32
25     18.28
26      8.04
27      1.10
28     17.89
29     14.16
30     15.56
31      5.42
32     11.53
33     15.76
34     11.49
35     13.57
36      1.34
37     12.12
38     14.74
39      1.44
40      9.11
41     17.12
42     16.15
43     12.89
44     10.76
45     19.25
46      1.15
47      5.51
48      8.52
49      2.17
50      4.79
51     12.26
52     14.83
53      2.12
54     11.53
55     10.76
56      1.44
57      2.60
58      2.65
59      4.06
60     13.91
61     14.40
62      1.63
63      6.73
64      3.33
65      3.72
66     10.61
67     11.92
68      4.74
69     10.51
70      5.85
71      2.21
72     12.94
73     16.92
74      1.58
75     14.69
76     17.70

0       1.19
1      20.17
2       0.00
3       0.00
4       9.74
5       3.23
6       0.00
7       2.75
8      12.21
9      14.01
10      0.00
11      6.63
12      0.00
13     20.03
14      6.68
15      6.87
16      0.00
17     11.19
18      0.00
19      0.00
20      0.00
21      0.00
22      2.60
23      6.68
24     14.83
25     12.31
26      0.00
27      1.83
28      0.00
29     12.89
30     14.98
31      0.00
32     16.15
33      0.00
34     16.87
35      0.00
36     16.78
37      0.00
38     13.96
39      0.00
40      2.94
41      0.00
42      0.00
43      3.38
44     17.99
45      2.50
46     16.05
47      2.02
48      7.17
49      2.07
50      3.82
51      0.00
52      0.00
53      1.83
54      0.00
55      0.00
56      1.34
57      0.00
58      2.50
59      0.00
60     17.17
61      4.54
62      0.00
63      6.68
64     10.32
65      0.00
66      9.01
67     14.11
68      0.00
69      0.00
70      3.52
71      9.88
72      7.36
73      0.00
74      8.91
75      0.00
76     18.77

0      18.82
1      20.03
2       0.00
3      18.57
4      16.68
5       5.51
6       0.00
7       7.60
8      15.66
9      17.50
10      0.00
11      7.21
12      0.00
13     18.23
14      1.53
15     18.04
16     14.59
17     13.52
18      0.00
19     10.42
20      0.00
21      8.67
22     13.52
23     10.42
24     11.34
25     15.51
26      7.55
27     12.02
28     13.04
29     11.39
30     15.17
31     16.00
32     12.17
33      0.00
34     13.48
35      0.00
36     18.38
37      0.00
38     13.96
39     10.08
40      8.48
41      6.39
42      0.00
43      7.12
44     17.94
45     13.28
46     16.78
47      7.12
48     11.39
49      3.38
50     10.85
51     13.48
52      0.00
53     17.02
54     20.03
55      0.00
56      7.99
57     12.84
58     17.65
59     20.22
60     16.19
61     16.97
62     19.59
63      7.65
64     13.43
65      4.25
66      9.40
67     14.11
68     18.23
69     11.83
70     13.28
71      6.39
72      6.63
73     20.32
74     15.42
75      4.40
76     18.52

0       1.92
1      19.98
2       0.00
3       0.00
4       2.26
5       9.11
6       0.00
7      11.53
8       4.01
9      15.27
10      0.00
11      5.85
12     15.85
13     20.17
14      6.19
15      6.63
16      0.00
17     18.67
18      0.00
19      0.00
20      0.00
21      0.00
22     13.62
23      7.02
24     12.70
25      0.00
26      5.95
27     11.05
28      0.00
29      0.00
30     13.72
31      0.00
32      8.86
33      0.00
34     16.87
35     18.14
36     13.33
37      0.00
38     13.96
39      0.00
40      7.17
41      0.00
42      0.00
43      3.86
44     13.09
45      2.75
46      1.63
47      1.73
48      4.83
49      2.46
50     10.42
51     13.33
52      0.00
53     13.38
54      0.00
55      0.00
56      7.41
57      0.00
58      1.92
59      1.73
60     16.15
61      9.20
62      0.00
63      5.27
64      6.44
65      0.00
66      9.40
67     14.40
68     18.38
69      0.00
70      5.61
71      9.74
72      3.09
73      0.00
74      4.01
75      0.00
76      2.70

0      19.30
1       9.69
2      18.33
3      18.67
4      18.82
5       7.55
6      20.22
7       1.97
8      12.84
9      13.52
10      0.00
11     10.51
12     18.04
13     19.16
14      9.40
15     12.80
16      8.23
17     16.10
18     11.39
19      9.06
20      9.06
21      3.82
22     16.73
23     14.98
24     13.48
25     12.26
26     15.61
27      7.41
28     12.75
29     14.88
30     19.79
31     11.78
32     13.77
33      7.80
34     18.43
35     17.84
36     10.17
37     13.52
38     14.16
39      9.45
40     17.02
41     18.18
42      0.00
43     18.86
44     19.11
45     10.51
46     19.83
47     17.12
48     12.02
49     12.84
50     16.87
51     14.50
52      0.00
53      3.86
54      7.12
55     10.27
56     16.73
57     10.61
58     10.13
59      6.58
60     19.06
61     13.57
62      0.00
63     14.88
64     15.76
65     19.98
66     12.55
67     13.57
68      3.23
69     16.00
70     17.12
71     16.68
72      9.06
73     20.22
74     16.63
75      7.70
76      3.14

0       4.16
1       1.97
2       0.00
3       0.00
4      15.71
5       6.39
6       0.00
7       8.09
8      12.41
9      17.80
10      0.00
11     12.46
12     14.98
13      3.86
14     19.64
15      6.49
16     13.33
17     16.49
18      0.00
19      8.52
20     11.05
21      0.00
22     19.88
23     10.90
24      5.95
25     14.93
26     15.08
27      9.40
28     13.96
29      0.00
30      9.01
31      0.00
32      7.36
33      0.00
34     12.12
35     16.92
36     16.73
37      0.00
38     14.16
39      0.00
40      2.50
41      9.25
42     11.49
43     10.08
44     19.25
45      5.61
46     17.84
47     10.32
48     12.65
49     19.69
50     11.78
51     13.72
52      0.00
53      4.93
54      0.00
55      6.78
56     12.65
57     10.51
58      2.75
59     14.35
60     16.29
61     14.64
62     16.39
63      2.07
64     10.71
65      6.39
66      9.25
67     12.65
68      3.86
69     13.14
70     13.09
71      8.67
72     10.47
73     19.83
74     11.44
75     19.98
76      3.14

0       0.00
1       6.00
2       0.00
3       0.00
4       0.00
5       1.34
6       0.00
7       1.10
8       6.24
9       0.00
10      0.00
11      8.09
12      0.00
13      3.43
14     14.54
15     15.22
16      0.00
17     11.58
18      0.00
19      0.00
20      0.00
21      9.74
22      0.00
23     17.65
24      0.00
25      0.00
26      0.00
27      0.00
28      0.00
29      0.00
30     10.66
31      0.00
32     13.77
33      0.00
34     11.39
35      0.00
36      0.00
37      0.00
38     13.82
39     14.59
40      0.00
41      0.00
42      4.64
43      7.46
44     14.20
45      0.00
46      0.00
47      0.00
48      0.00
49     16.92
50     12.21
51      7.31
52      2.07
53      3.62
54      0.00
55      0.00
56      4.50
57      0.00
58      2.99
59      0.00
60      0.00
61      4.93
62      0.00
63      7.99
64      0.00
65      0.00
66      0.00
67      0.00
68      0.00
69      0.00
70      2.12
71      0.00
72      3.77
73      0.00
74      2.84
75      0.00
76     18.43

0       0.00
1      14.93
2       0.00
3       0.00
4      16.24
5       4.79
6       0.00
7      14.54
8      10.71
9       0.00
10      0.00
11     12.46
12      0.00
13     19.88
14     17.07
15      4.40
16      0.00
17     18.33
18      9.45
19      0.00
20      0.00
21      0.00
22      0.00
23     18.43
24      0.00
25      0.00
26      0.00
27      0.00
28      0.00
29      0.00
30     13.14
31      0.00
32     12.02
33      0.00
34     18.72
35      0.00
36      0.00
37      0.00
38     13.82
39      0.00
40      0.00
41      0.00
42      0.00
43      9.50
44     13.77
45     18.52
46      0.00
47      0.00
48      0.00
49      2.07
50     14.64
51      0.00
52      0.00
53     15.47
54      0.00
55      6.58
56     16.68
57      0.00
58      2.60
59      0.00
60      0.00
61      6.87
62      0.00
63     11.15
64      0.00
65      0.00
66      0.00
67      0.00
68      0.00
69      0.00
70     10.56
71      0.00
72      4.74
73      0.00
74     15.22
75      1.24
76      4.11

0      13.82
1      13.82
2       0.00
3       0.00
4      16.19
5       7.46
6       0.00
7      18.23
8      15.37
9       0.00
10      0.00
11     12.46
12      0.00
13     19.11
14     13.72
15     14.11
16      0.00
17     11.83
18      0.00
19      0.00
20      0.00
21      0.00
22      0.00
23     14.06
24      0.00
25      0.00
26      0.00
27      0.00
28     11.24
29      0.00
30     10.17
31      0.00
32     12.07
33      0.00
34     10.47
35      0.00
36      0.00
37      0.00
38     14.74
39      0.00
40      0.00
41      0.00
42      0.00
43      8.91
44     14.64
45     17.75
46      0.00
47      9.40
48      0.00
49     20.08
50     17.55
51      0.00
52      0.00
53     10.56
54      0.00
55      0.00
56      5.17
57      0.00
58      2.36
59      0.00
60      0.00
61      2.99
62      0.00
63      7.89
64      0.00
65      0.00
66      0.00
67      0.00
68      0.00
69      0.00
70     12.60
71      9.11
72      9.98
73      0.00
74     17.17
75      0.00
76     19.20

In [38]:
for index, rows in ratings.iterrows():
    print index

0
1
2
3
4
5
6
7
8
9
10
11
12
13
14
15
16
17
18
19
20
21
22
23
24
25
26
27
28
29
30
31
32
33
34
35
36
37
38
39
40
41
42
43
44
45
46
47
48
49
50
51
52
53
54
55
56
57
58
59
60
61
62
63
64
65
66
67
68
69
70
71
72
73
74
75
76
77
78
79
80
81
82
83
84
85
86
87
88
89
90
91
92
93
94
95
96
97
98
99
100
101
102
103
104
105
106
107
108
109
110
111
112
113
114
115
116
117
118
119
120
121
122
123
124
125
126
127
128
129
130
131
132
133
134
135
136
137
138
139
140
141
142
143
144
145
146
147
148
149
150
151
152
153
154
155
156
157
158
159
160
161
162
163
164
165
166
167
168
169
170
171
172
173
174
175
176
177
178
179
180
181
182
183
184
185
186
187
188
189
190
191
192
193
194
195
196
197
198
199
200
201
202
203
204
205
206
207
208
209
210
211
212
213
214
215
216
217
218
219
220
221
222
223
224
225
226
227
228
229
230
231
232
233
234
235
236
237
238
239
240
241
242
243
244
245
246
247
248
249
250
251
252
253
254
255
256
257
258
259
260
261
262
263
264
265
266
267
268
269
270
271
272
273
274
275
276
27

In [84]:
import itemBasedRec as ibr
reload(ibr)

<module 'itemBasedRec' from 'itemBasedRec.py'>

In [54]:
ibr.test(ratings, 0.2, 'standEst')

KeyError: (1, 0)

In [61]:
ibr.cross_validate_user(ratings, 10, 0.2, estMethod=ibr.standEst, simMeas=ibr.pearsSim)

KeyError: (10, 0)

In [63]:
dataMat = np.genfromtxt('modified_jester_data.csv',delimiter=',')

In [85]:
ibr.cross_validate_user(dataMat, 10, 0.2, estMethod=ibr.standEst, simMeas=ibr.pearsSim)

(25.911595766755333, 7)

In [68]:
test_indices = np.random.randint(0, 100, 2)

In [69]:
test_indices

array([19, 98])

In [74]:
number_of_items = np.shape(dataMat)[1]

In [75]:
number_of_items

100L

In [76]:
rated_items_by_user = np.array([i for i in range(number_of_items) if dataMat[1,i]>0])

In [77]:
rated_items_by_user

array([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23, 24, 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 40, 41, 42, 43, 44, 45, 46, 47, 48, 49, 50, 51, 52, 53, 54, 55, 56, 57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 67, 68, 69, 70, 71, 72, 73, 74, 75, 76, 77, 78, 79, 80, 81, 82, 83, 84, 85, 86, 87, 88, 89, 90, 91, 92, 93, 94, 95, 96, 97, 98, 99])

In [78]:
test_size = 0.2 * len(rated_items_by_user)

In [79]:
test_size

20.0

In [82]:
test_indices = np.random.randint(0, len(rated_items_by_user), int(test_size)) # errors here